### PR TITLE
feat(hogql): Allow lazy joins on lazy tables with requested fields

### PR DESCRIPTION
**Pull request by Gilbert09**
## Problem
- When adding a `LazyJoin` onto `persons` that uses any constraint field other than `person.id`, the `lazy_tables` resolver doesn't include the constraint fields into the `persons` `LazyTable` subquery, and thus causes the lazy join to fail field resolution

## Changes
- If the lazy join uses fields on the constraint other than whats already pre-defined in the lazy tables select subquery, then pass these into the subquery
- ~Passes in an alias to the subquery to cater for selecting the likes of `properties.field`, and update the same alias in the join function constraint~
  - I wanted this to happen from within `lazy_tables`, but I've pushed this functionality down to the individual `join_function` instead. This isn't ideal, as each new lazy join may need to implement the same logic, but the complexity of finding the right `ast.Field`'s and their associated types from within `lazy_tables` to update felt hackier and more prone to error
- Fixed a bunch of mypy errors within `lazy_tables`
- Added UI to the data warehouse join modal to allow for hogql expressions

**Edit:**
- Updated lazy tables to support `TableAliasType` nodes with a child of either `LazyJoinType` or `LazyTableType` - turns out lazy tables just wasn't running on anything that was aliased before, god knows how this managed to run without erroring (or maybe it just returned bad data..)
- Upgraded the existing logic that tells lazy joins which fields to request - this now also rewrites the constraint field chains so that property types can be used within constraints, and then removed the override logic in the individual lazy join function that was added earlier
- `LazyJoin` `from_field` field now supports full field chains (e.g. `[properties.email]`), and have added an optional `to_field` to help support the data warehouse joins - this is only used in the override logic and is only needed when a dot notated field is used on the right side of the join constraint

## How did you test this code?
- Added clickhouse snapshotting for lazy tables to hand verify the queries are valid

### CODE REVIEW

File: posthog/hogql/transforms/test/test_lazy_tables.py [hunk 1]
```diff
@@ -86,3 +87,47 @@ def _print_select(self, select: str):
...
+            source_table_name="persons",
+            source_table_key="$hogql",
+            source_table_key_hogql="properties.email",
+            joining_table_name="events",
+            joining_table_key="event",
+            field_name="events",
+        ).save()
+
+        printed = self._print_select("select events.event from persons")
+        assert printed == self.snapshot
```
Comment (mariusandra) [inline posthog/hogql/transforms/test/test_lazy_tables.py:133]: This example seems more artificial than the others. 🤔

Comment (mariusandra) [reply posthog/hogql/transforms/test/test_lazy_tables.py:133]: Also, might be good to test these tests with both materialized and non-materialized properties.

File: posthog/hogql/transforms/lazy_tables.py [hunk 2]
```diff
@@ -203,13 +251,34 @@ def visit_select_query(self, node: ast.SelectQuery):
...
+                and len(join_to_add.table.chain) > 0
+            ):
+                join_to_add_table_name = str(join_to_add.table.chain[0])
+
+            if (
+                join_to_add_table_name is not None
+                and node.type is not None
+                and node.type.tables.get(join_to_add_table_name) is not None
+            ):
+                node.type.tables.pop(join_to_add_table_name)
```
Comment (mariusandra) [inline posthog/hogql/transforms/lazy_tables.py:275]: Could this cause problems in some case? Is there any time when a legit join with the same name could have been done to a table?

 E.g.
```sql
   select event, person.properties.email as real_email
     from events 
left join person as person 
       on (person.properties.email = events.properties.tr00_mail)
left join person as person2
       on (person2.properties.email = events.properties.fake_mail)
```

I haven't followed the code to verify if this would be an issue or not, but would this run?

**Review by mariusandra — COMMENTED** [review]
Overall looks good. I did add some 🤔  inline, and I'm also wondering if there's a reason `source_table_key_hogql` and `joining_table_key_hogql` can't just be rolled into `source_table_key` and `source_table_key`?

File: posthog/hogql/transforms/lazy_tables.py (hunk 2 above)
Comment (Gilbert09) [reply posthog/hogql/transforms/lazy_tables.py:275]: I believe in this example, this would be done on the alias and not the table itself, (e.g. `person` and `person2`). But I'll test this to verify

**Comment by Gilbert09** [comment]
@mariusandra Gonna re-request your review, managed to redo this without having to resolve types twice etc, feels nicer now and we even support table aliases 😱

File: frontend/src/scenes/data-warehouse/viewLinkLogic.tsx [hunk 3]
```diff
@@ -112,44 +164,23 @@ export const viewLinkLogic = kea<viewLinkLogicType>([
...
                     try {
                         await api.dataWarehouseViewLinks.update(values.joinToEdit.id, {
                             source_table_name: source_table_name ?? values.selectedSourceTable.name,
-                            source_table_key,
+                            source_table_key: values.selectedSourceKey ?? undefined,
+                            source_table_key_hogql: values.selectedSourceKeyHogQL ?? undefined,
                             joining_table_name,
-                            joining_table_key,
+                            joining_table_key: values.selectedJoiningKey ?? undefined,
+                            joining_table_key_hogql: values.selectedJoiningKeyHogQL ?? undefined,
```
Comment (EDsCODE) [inline frontend/src/scenes/data-warehouse/viewLinkLogic.tsx:183]: Maybe can reuse the non hogql join key?

File: posthog/hogql/transforms/lazy_tables.py [hunk 4]
```diff
@@ -62,7 +96,12 @@ def visit_property_type(self, node: ast.PropertyType):
                 self.stack_of_fields[-1].append(node)
 
     def visit_field_type(self, node: ast.FieldType):
-        if isinstance(node.table_type, ast.LazyJoinType) or isinstance(node.table_type, ast.LazyTableType):
+        if isinstance(node.table_type, ast.TableAliasType):
```
Comment (EDsCODE) [inline posthog/hogql/transforms/lazy_tables.py:99]: yikes

File: posthog/hogql/transforms/lazy_tables.py [hunk 5]
```diff
@@ -21,18 +21,48 @@ def resolve_lazy_tables(
...
+class FieldChainReplacer(TraversingVisitor):
+    overrides: List[ConstraintOverride] = {}
+
+    def __init__(self, overrides: List[ConstraintOverride]) -> None:
+        super().__init__()
+        self.overrides = overrides
+
+    def visit_field(self, node: ast.Field):
+        for constraint in self.overrides:
+            if node.chain == constraint.chain_to_replace:
```
Comment (EDsCODE) [inline posthog/hogql/transforms/lazy_tables.py:52]: If this is traversing the entire query again, is it possible that the chain_to_replace matches something unintentionally? It's a super nit edge case but if an outer query reuses an alias/table name that exists on an inner query, it could get overwritten by this override?

**Review by EDsCODE — APPROVED** [review]
Some comments but overall makes sense. Good catch on the aliasing

File: frontend/src/scenes/data-warehouse/viewLinkLogic.tsx (hunk 3 above)
Comment (Gilbert09) [reply frontend/src/scenes/data-warehouse/viewLinkLogic.tsx:183]: I had thought about not being able to differentiate between picking a field on a table vs a hogql expression, but I've just realized that we can do `parse_expr(selectedJoiningKey)` and it'll return a Field with a single chain, which is what we want. I'll simplify this now

File: frontend/__snapshots__/scenes-app-insights--user-paths-edit--light--webkit.png
Comment (mariusandra) [inline frontend/__snapshots__/scenes-app-insights--user-paths-edit--light--webkit.png:1]: This change is concerning

File: posthog/hogql/transforms/lazy_tables.py [hunk 6]
```diff
@@ -252,3 +394,12 @@ def visit_select_query(self, node: ast.SelectQuery):
...
         self.stack_of_fields.pop()
+
+        # When joining a lazy table to another lazy table, the joined table doesn't get resolved
+        # Doing another pass solves this for us
+        if self.lazy_finder_counter < 20:
+            lazy_finder = LazyFinder()
+            lazy_finder.visit(node)
+            if lazy_finder.found_lazy:
+                self.lazy_finder_counter = self.lazy_finder_counter + 1
+                self.visit_select_query(node)
```
Comment (mariusandra) [inline posthog/hogql/transforms/lazy_tables.py:405]: This feels like code nobody will dare touch for years 😅. I wonder if there's a smarter way to just flag tables that should be revisited, instead of doing it for everything?

**Review by mariusandra — APPROVED** [review]
It got hard to follow in the `lazy_table.py` part, but I'm going to assume all is good and/or you'll fix it if not 😅.
